### ISSUE TITLE
Cross Validation

### ISSUE THREAD

**Issue by magicmutal** [ageron/handson-ml#157]
Hi, my name is Jinsu.
Your book is very helpful to me, thank you.

My issue is in page 71, 

"Now the "Decision Tree" doesn't look as good as it did earlier. In fact, it seems to perform worse than the Linear Regression model"

how do you come to this conclusion??

I can understand that "lin_rmse" is larger than "tree_rmse"  so it looks like "Decision Tree" performs better., but I cannot understand how do I say "Decision Tree" perform worse.

What is the meaning of "scores"??

**Comment by ageron** [comment]
Hi Jinsu,

Thanks for your question.  The first time we tried the `DecisionTree`, we got perfect predictions:

```python
>>> housing_predictions = tree_reg.predict(housing_prepared)
>>> tree_mse = mean_squared_error(housing_labels, housing_predictions)
>>> tree_rmse = np.sqrt(tree_mse)
>>> tree_rmse
0.0
```

RMSE equal to 0, means zero errors, perfect predictions. But this was because we were evaluating the model on the training set, and we had severely overfitted the training set (like learning it by heart, including all the noise).

When we use cross-validation instead we see the "true" performance of the decision tree model:

```python
from sklearn.model_selection import cross_val_score
scores = cross_val_score(tree_reg, housing_prepared, housing_labels,
                         scoring="neg_mean_squared_error", cv=10)
tree_rmse_scores = np.sqrt(-scores)
```

And later:
```python

>>> display_scores(tree_rmse_scores)
Scores: [ 70232.0136482   66828.46839892  72444.08721003  70761.50186201
  71125.52697653  75581.29319857  70169.59286164  70055.37863456
  75370.49116773  71222.39081244]
Mean: 71379.0744771
Standard deviation: 2458.31882043
```

The training set was split into 10 chunks. The model was first trained on all chunks except chunk 1, and evaluated on chunk 1: this gives the first score (RMSE). Then it was trained again on all chunks except chunk 2, and evaluated on chunk 2: this gives the second score, and so on. At the end, we can compute the mean score, and we see 71379.0744771: this is a much better evaluation of the performance of the model, because it is based only on predictions made on instances that the model did not see during training.  In this case, it is a pretty large error, so it is clear that even though the model has zero error on the training set, it is actually a bad model because it does not generalize well at all to instances it has never seen.

I said that it performs even worse than the linear model, which is true, but to be fair we should look at the linear model's performance using cross-validation as well, because it may be possible that it is overfitting the training set a bit.  However, linear models are so simple that they usually don't overfit the training set much, unless there are very few training instances.

I hope this helps,
Aurélien
